logsrvd/logsrvd.c

```diff
@@ -784,6 +825,309 @@ signal_cb(int signo, int what, void *v)
+ * be able to establish configured network connections over TLS
+ */
+static SSL_CTX *
+init_tls_server_context(void)
+{
+    const SSL_METHOD *method;
+    SSL_CTX *ctx = NULL;
+    const struct logsrvd_tls_config *tls_config = logsrvd_get_tls_config();
+
+    debug_decl(init_tls_server_context, SUDO_DEBUG_UTIL)
+
+    SSL_library_init();
+    OpenSSL_add_all_algorithms();
+    SSL_load_error_strings();
+
+    if ((method = TLS_server_method()) == NULL) {
+        sudo_debug_printf(SUDO_DEBUG_ERROR|SUDO_DEBUG_LINENO,
+            "creation of SSL_METHOD failed: %s",
+            ERR_error_string(ERR_get_error(), NULL));
+        goto bad;
+    }
+    if ((ctx = SSL_CTX_new(method)) == NULL) {
+        sudo_debug_printf(SUDO_DEBUG_ERROR|SUDO_DEBUG_LINENO,
+            "creation of new SSL_CTX object failed: %s",
+            ERR_error_string(ERR_get_error(), NULL));
+        goto bad;
+    }
+
+    /* verify server certification against the CA bundle file */
+    if (!verify_server_cert(ctx, tls_config)) {
+        goto bad;
+    }
+
+    /* if peer authentication is enabled, verify client cert during TLS handshake */
+    if (tls_config->check_peer) {
+        X509 *cacert = load_cert(tls_config->cacert_path);
+
+        /* server will send the name of the CA to the client during the handshake */
+        if (SSL_CTX_add_client_CA(ctx, cacert) <= 0) {
+            sudo_debug_printf(SUDO_DEBUG_ERROR|SUDO_DEBUG_LINENO,
+                "calling SSL_CTX_add_client_CA() failed: %s",
+                ERR_error_string(ERR_get_error(), NULL));
+        }
+
+        /* sets the location of the CA bundle file for verification purposes */
+        if (SSL_CTX_load_verify_locations(ctx, tls_config->cacert_path, NULL) <= 0) {
+            sudo_debug_printf(SUDO_DEBUG_ERROR|SUDO_DEBUG_LINENO,
+                "calling SSL_CTX_load_verify_locations() failed: %s",
+                ERR_error_string(ERR_get_error(), NULL));
+        }
+
+        /* server will send a client certificate request to the peer during the handshake */
+        SSL_CTX_set_verify(ctx, SSL_VERIFY_PEER | SSL_VERIFY_FAIL_IF_NO_PEER_CERT, NULL);
+    }
+
+    if (!SSL_CTX_use_PrivateKey_file(ctx, tls_config->pkey_path, SSL_FILETYPE_PEM)) {
+        sudo_debug_printf(SUDO_DEBUG_ERROR|SUDO_DEBUG_LINENO,
+            "unable to load key file: %s",
+            ERR_error_string(ERR_get_error(), NULL));
+        goto bad;
+    }
+    if (!SSL_CTX_check_private_key(ctx)) {
+        sudo_debug_printf(SUDO_DEBUG_ERROR|SUDO_DEBUG_LINENO,
+            "unable to verify key file: %s",
+            ERR_error_string(ERR_get_error(), NULL));
+        goto bad;
+    }
+
+    /* initialize TLSv1.2 and TLSv1.3 ciphersuites */
+    if (!init_tls_ciphersuites(ctx, tls_config)) {
+        goto bad;
+    }
+
+    /* try to load and set diffie-hellman parameters  */
+    FILE *dhparam_file = fopen(tls_config->dhparams_path, "r");
+    if (dhparam_file) {
```

Comment:
For pointers I'd prefer that they be compared to NULL rather than use implicit boolean checks

Reply:
fixed